ext/drx/drx.c

```diff
@@ -556,6 +558,7 @@ drx_insert_counter_update(void *drcontext, instrlist_t *ilist, instr_t *where,
      * address being near this one, and add to reg1 instead of
      * taking 2 instrs to load it fresh.
      */
+    /* update counter */
```

Comment:
Hmm, it seems like for our use case at least we would want release-acquire semantics to ensure writes are visible promptly and in order to other threads, even though we're ok not having a pure atomic fetch-and-add and thus having some races.  I think the existing 32-bit arm code would want that too.  I would vote for just putting in the release-acquire, though you could argue for a 3rd API mode to let clients use completely relaxed models.  This makes the racy counter update behave like the x86 one, which at least matches my intuition for how it should operate.

So I'm voting for `ldar` and `stlr` for a64 and `dmb ish;str` and `ldr;dmb ish` for a32, and mentioning the release-acquire in the docs.

Reply:
Or are there common cases of clients incrementing thread-private values and they don't want to pay for barriers?  Maybe get some more opinions: @johnfxgalea?

Reply:
I could see cases where one does not need barriers for thread-private counters, but given the simplicity of increments, one can easily go around this problem by not using this API. Besides, the 3rd API mode that you suggested earlier might also accept a flag to avoid barriers if we need this in the future.

Reply:
Maybe add a new flag as DRX_COUNTER_LOCK is for atomic behaviour.

Reply:
as suggested, updated counter updates for 32-bit and 64-bit arm with release-acquire semantics and added a flag for optional completely relaxed model. Also, updated the description in the header with info about the new flag.